Samples/src/OrderFinding/Program.cs

```diff
@@ -0,0 +1,158 @@
+using System.Linq;
+
+namespace Microsoft.Quantum.Samples.OrderFinding
+{
+    /// <summary>
+    /// This class holds a permutation π and allows to compute its
+    /// order in various different ways (exact, guess classically,
+    /// guess with quantum)
+    /// </summary>
+    class Permutation
+    {
+        /// <summary>
+        /// The permutation object is constructed with a permutation π and
+        /// a number of shots for repetitions when guessing.
+        /// <param name="permutation">Permutation π over 4 elements 0, 1, 2, 3</param>
+        /// <param name="shots">Number of repetitions when guessing</param>
+        /// </summary>
+        public Permutation(List<int> permutation, int shots = 1024)
+        {
+            _p = permutation;
+            _shots = shots;
+        }
+
+        /// <summary>
+        /// Returns the exact order (length) of the cycle that contains given index
+        /// <summary>
+        public int ComputeOrder(int index)
+        {
+            int order = 1, cur = index;
+            while (index != _p[cur])
+            {
+                ++order;
+                cur = _p[cur];
+            }
+            return order;
+        }
+
+        /// <summary>
+        /// Guesses the order (classically) for cycle that contains index
+        ///
+        /// The algorithm computes π³(index).  If the result is index, it
+        /// returns 1 or 3 with probability 50% each, otherwise, it
+        /// returns 2 or 4 with probability 50% each. 
+        /// <summary>
+        private int GuessOrderClassicalOne(int index)
+        {
+            if (_p[_p[_p[index]]] == index)
+            {
+                return _rnd.Next(2) == 0 ? 1 : 3;
+            }
+            else
+            {
+                return _rnd.Next(2) == 0 ? 2 : 4;
+            }
+        }
+
+        /// <summary>
+        /// Guesses order classically for shots times, and returns the percentage for each order that was returned.
+        /// </summary>
+        public IEnumerable<(int order, double percentage)> GuessOrderClassical(int index)
+        {
+            return Enumerable.Range(0, _shots).Select(_ => GuessOrderClassicalOne(index)).GroupBy(n => n, (n, list) => (n, list.Count() / (double)_shots));
+        }
+
+        /// <summary>
+        /// The quantum estimation calls the quantum algorithm in the Q# file which computes the permutation
+        /// πⁱ(input) where i is a superposition of all values from 0 to 7.  The algorithm then uses QFT to
+        /// find a period in the resulting state.  The result needs to be post-processed to find the estimate.
+        /// <summary>
+        private int GuessOrderQuantumOne(int index)
+        {
+            var result = OrderFinding.Run(_sim, new QArray<long>(_p.ConvertAll(x => (long)x)), (long)index).Result;
+
+            if (result == 0)
+            {
+                var guess = _rnd.NextDouble();
+                if (guess <= 0.5505)
```

Comment:
please explain magic constants for the branchings. Perhaps put a pointer to the original paper.

Reply:
Fixed in 3e0ecdd.